### PR TITLE
Prevent windows msys shell from interpreting @{upstream} arg

### LINKED ISSUE #737: Windows can not handle @{upstream} correcetly
```
W1116 14:16:38.877539    8220 update.go:168] failed to update index "default": reset index at "C:\\Users\\mattn\\.krew\\index\\default" failed: command execution failure, output="fatal: ambiguous argument '@upstream': unknown revision or path not in the working tree.\nUse '--' to separate paths from revisions, like this:\n'git <command> [<revision>...] -- [<file>...]'\n": exit status 128
failed to update the following indexes: default
: reset index at "C:\\Users\\mattn\\.krew\\index\\default" failed: command execution failure, output="fatal: ambiguous argument '@upstream': unknown revision or path not in the working tree.\nUse '--' to separate paths from revisions, like this:\n'git <command> [<revision>...] -- [<file>...]'\n": exit status 128
```

### PR DESCRIPTION
**Pull request by mattn**
Fixes #737

### CODE REVIEW

**Comment by ahmetb** [comment]
Thanks for the patch. I'd love to understand this first.

To me this is very awkward because you'd think something like this would be necessary only if we were executing the `git` process in a `/bin/bash -c "..."` or `cmd.exe /c "..."`.

but we're doing this execution through the Go os/exec package which uses [CreateProcess](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-createprocessa?redirectedfrom=MSDN) API in Win32, that's why I am not fully getting why such escaping is necessary.

Does this syscall somehow have an arguments interpretation/evaluation step that we're not aware of? It would be good to learn why's this happening first.

**Comment by mattn** [comment]
msys2 shell handle `@{}` in peculiar argument parser.

image

**Comment by mattn** [comment]
image

**Comment by mattn** [comment]
FYI, On Windows, arguments are parsed in the process it-self. (not a shell part like UNIX)

**Comment by ahmetb** [comment]
@mattn I still don't understand the problem and I do need to understand the problem fully so that we do not break other hundreds or thousands of Windows users who have been using Krew on Windows just fine.

So it seems in your gcc example that the args are sent to the process correctly. (I don't know why you're showing msys echo.exe here).

What matters here is whether `git` is handling the `@{upstream}` arg correctly and based on other Windows users I have asked, the answer for them seems to be yes, that argument works just fine. I want to understand why is your git behaving differently.

**Comment by mattn** [comment]
If the program is linked with msys2 or cygwin runtime, that escaping is required. AFAIK, git on Windows shound be linked those runtimes.

**Comment by mattn** [comment]
One another candidate for fixing is quote.
image

**Comment by mattn** [comment]
This is a part converting that `@{}`.

https://github.com/cygwin/cygwin/blob/44a79a6eca3d322020dda0919023d78dda129d4d/winsup/cygwin/dcrt0.cc#L338-L344

**Comment by mattn** [comment]
How about to use `remotes/origin/HEAD` instead of `@{upstream}` ?

**Comment by mattn** [comment]
If found a spell to fix this issue `MSYS=noglob`.

File: internal/gitutil/git.go [hunk 1]
```diff
@@ -78,6 +79,9 @@ func GetRemoteURL(dir string) (string, error) {
 func Exec(pwd string, args ...string) (string, error) {
 	klog.V(4).Infof("Going to run git %s", strings.Join(args, " "))
 	cmd := osexec.Command("git", args...)
+	if runtime.GOOS == "windows" {
+		cmd.Env = append(os.Environ(), "MSYS=noglob")
```
Comment (mattn) [inline internal/gitutil/git.go:83]: If you prefer to append CYGWI=noglob also, please let me known.

Or we should append the variable like below?
```
msys := os.Getenv("MSYS") + ":noglob"
```

**Comment by ahmetb** [comment]
> How about to use `remotes/origin/HEAD` instead of `@{upstream}` ?

IIRC this was previously discussed and we wanted to do it this way for a reason, but I might be wrong. See the blame history, maybe that helps.



> I found a spell to fix this issue `MSYS=noglob`.

That's great, although again, I am not sure if _we_ should set this or whether it should be set by the user in their .bashrc or whatever they use. I am not sure what % of Krew users are using MSYS/Cygwin, but this little env-override (especially since it's only for windows, and only does _something_ on cygwin/msys) might be okay...

**Comment by mattn** [comment]
I think this is less wrong for users for msys2 git.

* This change only affect to users who use msys2 git.
* This change only affect to this git command.

File: internal/gitutil/git.go [hunk 2]
```diff
@@ -78,6 +79,26 @@ func GetRemoteURL(dir string) (string, error) {
 func Exec(pwd string, args ...string) (string, error) {
 	klog.V(4).Infof("Going to run git %s", strings.Join(args, " "))
 	cmd := osexec.Command("git", args...)
+
+	// NOTE: https://github.com/kubernetes-sigs/krew/pull/739
+	if runtime.GOOS == "windows" {
+		environ := os.Environ()
```
Comment (ahmetb) [inline internal/gitutil/git.go:85]: do you mind if we move `environ:=...` and `cmd.Env=environ` to outside the `if`?

File: internal/gitutil/git.go [hunk 3]
```diff
@@ -78,6 +79,26 @@ func GetRemoteURL(dir string) (string, error) {
 func Exec(pwd string, args ...string) (string, error) {
 	klog.V(4).Infof("Going to run git %s", strings.Join(args, " "))
 	cmd := osexec.Command("git", args...)
+
+	// NOTE: https://github.com/kubernetes-sigs/krew/pull/739
```
Comment (ahmetb) [inline internal/gitutil/git.go:83]: let's add a bit of more explanation here, e.g.

// NOTE: on windows, msys2 and cygwin interpret args e.g. `@{upstream}` and therefore
// acts differently than git-for-windows. see https://github.com/kubernetes-sigs/krew/pull/739

Comment (mattn) [reply internal/gitutil/git.go]: Thank you.

File: internal/gitutil/git.go (hunk 2 above)
Comment (mattn) [reply internal/gitutil/git.go]: Sure.

**Comment by ahmetb** [comment]
/retitle Prevent windows msys shell from interpreting @{upstream} arg

**Comment by ahmetb** [comment]
for future reference, I'll drop the issue you opened 5 years ago here ;) https://github.com/msys2/MSYS2-packages/issues/522#issue-144797722

/cc @corneliusweig 
any concerns?

**Review by corneliusweig — APPROVED** [review]
Looks good to me. It's a bit sad that this setting needs to  be added in application code and not in a more central place. But I guess that  would break other msys/cygwin usage patterns.

Fun story: by chance I just saw that we transitively depend on your go-colorable project :)

**Comment by mattn** [comment]
> But I guess that would break other msys/cygwin usage patterns.

FYI, go-colorable does not break msys/cygwin.

image

**Comment by mattn** [comment]
Sorry, I was confused and commented wrong mistaken with another PR.

**Pull request closed without merging**